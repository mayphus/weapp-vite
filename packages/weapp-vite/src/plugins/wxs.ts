import type { CompilerContext } from '@/context'
import type { WxmlDep } from '@/types'
import type { EmittedFile, PluginContext } from 'rollup'
import type { Plugin } from 'vite'
import { jsExtensions } from '@/constants'
import { transformWxsCode } from '@/wxs'
import { removeExtension } from '@weapp-core/shared'
import fs from 'fs-extra'
import { LRUCache } from 'lru-cache'
import path from 'pathe'

export const wxsCodeCache = new LRUCache<string, string>(
  {
    max: 512,
  },
)

export function wxs({ configService, wxmlService }: CompilerContext): Plugin[] {
  const wxsMap = new Map<string, {
    emittedFile: EmittedFile
  }>()

  async function transformWxs(this: PluginContext, {
    wxsPath,
  }: {
    wxsPath: string
  }) {
    if (await fs.exists(wxsPath)) {
      this.addWatchFile(wxsPath)
      const arr = wxsPath.match(/\.wxs(\.[jt]s)?$/)
      let isRaw = true
      if (arr) {
        isRaw = !arr[1]
      }
      const rawCode = await fs.readFile(wxsPath, 'utf8')
      let code = wxsCodeCache.get(rawCode)

      const dirname = path.dirname(wxsPath)
      if (code === undefined) {
        const { result, importees } = transformWxsCode(rawCode, {
          filename: wxsPath,
        })
        if (typeof result?.code === 'string') {
          code = result.code
        }
        await Promise.all(
          importees.map(({ source }) => {
            return transformWxs.call(this, {
              wxsPath: path.resolve(dirname, source),
            })
          }),
        )
      }

      if (code !== undefined) {
        wxsMap.set(wxsPath, {
          emittedFile: {
            type: 'asset',
            fileName: configService.relativeAbsoluteSrcRoot(isRaw ? wxsPath : removeExtension(wxsPath)),
            source: code,
          },
        })
        wxsCodeCache.set(rawCode, code)
      }
    }
  }
  function handleWxsDeps(this: PluginContext, deps: WxmlDep[], absPath: string) {
    return Promise.all(
      deps.filter(x => x.tagName === 'wxs').map(async (wxsDep) => {
        const arr = wxsDep.value.match(/\.wxs(\.[jt]s)?$/)
        if (jsExtensions.includes(wxsDep.attrs.lang) || arr) {
          const wxsPath = path.resolve(path.dirname(absPath), wxsDep.value)

          await transformWxs.call(this, {
            wxsPath,
          })
        }
      }),
    )
  }

  return [
    {
      name: 'weapp-vite:wxs',
      enforce: 'pre',
      buildStart() {
        wxsMap.clear()
      },

      async buildEnd() {
        await Promise.all(wxmlService.tokenMap.entries().map(([id, token]) => {
          return handleWxsDeps.call(this, token.deps, id)
        }))

        wxsMap.values().forEach(({ emittedFile }) => {
          this.emitFile(emittedFile)
        })
      },
    },
  ]
}
