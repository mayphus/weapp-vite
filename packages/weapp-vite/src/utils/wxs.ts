import type { BuildOptions } from 'esbuild'
// import type { TsupOptions } from '../types'
import { defu } from '@weapp-core/shared'
import { build as esbuild } from 'esbuild'
// import { build } from 'vite'
// import { build } from 'vite'

export async function buildWxs(options: BuildOptions) {
  // build({
  //   build:{
  //     target:
  //   }
  // })
  // const { build: tsupBuild } = await import('tsup')
  // build({})
  const mergedOptions: BuildOptions = defu<BuildOptions, BuildOptions[]>(options, {
    // entry: wxsPaths,
    format: 'cjs',
    // outDir: ctx.outDir,
    target: 'es5',
    // silent: true,
    // shims: true,
    loader: {
      '.wxs': 'js',
    },
    entryNames: '[dir]/[name]',
    outExtension: {
      '.js': '.wxs',
    },
    sourcemap: false,
    // outbase: 'src',
    // esbuildOptions: (options) => {
    //   // options.entryNames
    //   // options.outExtension
    //   options.entryNames = '[dir]/[dir]/[name]'
    //   options.outExtension = {
    //     '.js': '.wxs',
    //   }
    //   options.assetNames = '[dir]/[name]'
    //   options.chunkNames = '[dir]/[name]'
    //   // 保持名称
    //   // options.keepNames = true
    // },
    // outExtension: ({ format, options }) => {
    //   // console.log(format)
    //   return {
    //     js: '.wxs', // .wxs',
    //   }
    // },
    // config: false,

  })
  await esbuild(mergedOptions)
}
