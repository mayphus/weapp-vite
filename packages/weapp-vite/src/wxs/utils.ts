import { addExtension, removeExtensionDeep } from '@weapp-core/shared'

export function normalizeWxsFilename(value: string) {
  return addExtension(removeExtensionDeep(value), '.wxs')
}
