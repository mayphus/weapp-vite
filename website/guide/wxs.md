# Wxs 增强

## 编程支持 (实验性)

`wxs` 做了一些实验性的语言支持，以至于除了使用 `index.wxs` 之外，你还能使用 `index.wxs.js` 和 `index.wxs.ts`

> `wxs.js` 和 `wxs.ts` 只能使用 `require` 语法，不能使用 `import` 语法

### 引入 wxs.js/wxs.ts

`wxs.js/wxs.ts` 里面可以看作是 `wxs` 的增强，你可以在里面使用一些进阶的语法

引入方式

```html
<wxs module="test" src="index.wxs.ts" />
```

这个在最终产物中会被处理为:

```html
<wxs module="test" src="index.wxs" />
```

同时对 `index.wxs.ts` 里面的内容进行编译，结果为 `wxs`

目前编译的能力还是比较弱的，因为 `wxs` 语法非常特殊，很多 `js` 里面是无法使用的，比如 `for ... in xxx` 这种基础语法。

### 内联 wxs

针对内联 `wxs` 做了一些特殊的优化，你只需在内联 `wxs` 代码块上声明 `lang` 是 `js` or `ts`，就会启动内置的转译引擎进行转化

比如下方的代码在 `weapp-vite` 里就是合法的

```html
<view>{{test.foo}}</view>
<view @tap="{{test.tigger}}">{{test.abc}}</view>

<wxs module="test" lang="ts">
const { bar, foo } = require('./index.wxs.js')
const bbc = require('./bbc.wxs')
export const abc = 'abc'

export function tigger(value:string){
  console.log(abc)
}

export {
  foo,
  bar,
  bbc
}
</wxs>
```
