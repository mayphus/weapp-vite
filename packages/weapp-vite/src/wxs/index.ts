import babel from '@babel/core'
import * as t from '@babel/types'
import { defu } from '@weapp-core/shared'
import { normalizeWxsFilename } from './utils'

export interface TransformWxsCodeOptions {
  filename?: string
  // inline?: boolean
}

export {
  normalizeWxsFilename,
}

export function transformWxsCode(code: string, options?: TransformWxsCodeOptions) {
  const { filename } = defu<TransformWxsCodeOptions, TransformWxsCodeOptions[]>(options, {
    filename: '_default.wxs',
  })
  return babel.transformSync(code, {
    presets: [
      ['@babel/preset-env'],
      ['@babel/preset-typescript'],
    ],
    parserOpts: {
      plugins: ['typescript'],
    },
    filename,
    plugins: [
      {
        visitor: {
          Directive: {
            enter(p) {
              p.remove()
            },
          },
          CallExpression: {
            enter(p) {
              if (p.get('callee').isIdentifier({
                name: 'require',
              }) && p.get('arguments').length === 1) {
                const importee = p.get('arguments')[0]
                if (importee.isStringLiteral()) {
                  importee.node.value = normalizeWxsFilename(importee.node.value)
                }
              }
            },
          },
          ExpressionStatement(p) {
            const expression = p.node.expression

            if (
              expression.type === 'CallExpression'
              && expression.callee.type === 'MemberExpression'
              && t.isIdentifier(expression.callee.object)
              && expression.callee.object.name === 'Object'
              && t.isIdentifier(expression.callee.property)
              && expression.callee.property.name === 'defineProperty'
              && expression.arguments.length >= 2
              && t.isIdentifier(expression.arguments[0])
              && expression.arguments[0].name === 'exports'
              && t.isStringLiteral(expression.arguments[1])
              && expression.arguments[1].value === '__esModule'
            ) {
              p.remove()
            }
          },
          NewExpression: {
            enter(p) {
              if (p.get('callee').isIdentifier({
                name: 'RegExp',
              })) {
                p.replaceWith(
                  t.callExpression(t.identifier('getRegExp'), p.get('arguments').map(x => x.node)),
                )
              }
              else if (p.get('callee').isIdentifier({
                name: 'Date',
              })) {
                p.replaceWith(
                  t.callExpression(t.identifier('getDate'), p.get('arguments').map(x => x.node)),
                )
              }
            },
          },
          RegExpLiteral: {
            enter(p) {
              const args = [t.stringLiteral(p.node.pattern)]
              if (p.node.flags) {
                args.push(t.stringLiteral(p.node.flags))
              }
              p.replaceWith(
                t.callExpression(t.identifier('getRegExp'), args),
              )
            },
          },
          MemberExpression: {
            enter(p) {
              if (p.get('object').isIdentifier({
                name: 'exports',
              })) {
                p.replaceWith(
                  t.memberExpression(
                    t.memberExpression(t.identifier('module'), t.identifier('exports')),
                    p.get('property').node,
                  ),
                )
              }
            },
          },
        },
      },
    ],
  })
}
