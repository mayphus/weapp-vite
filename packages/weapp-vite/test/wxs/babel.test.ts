import { transformWxsCode } from '@/wxs'
import fs from 'fs-extra'
import path from 'pathe'

describe('babel', () => {
  it('should 0', async () => {
    const res = await transformWxsCode(`export const foo = "'hello world' from comm.wxs";
export const bar = function (d: string) {
  return d;
}
`)

    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })

  it('should 1', async () => {
    const res = await transformWxsCode(`export const foo = "'hello world' from comm.wxs";
export const bar = function (d) {
  return d;
}
`)

    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })

  it('should 2', async () => {
    const res = await transformWxsCode(`export function aa() {
  return new RegExp('{|}|"', 'g')
}
`)

    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })

  it('should 3', async () => {
    const res = await transformWxsCode(`export function bb() {
  return /{|}|"/g
}
`)

    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })

  it('should 4', async () => {
    const res = await transformWxsCode(`export function bb() {
  return /{|}|"/
}
`)

    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })

  it('should 5', async () => {
    const res = await transformWxsCode(`export function aa() {
  return new RegExp('{|}|"')
}
`)

    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })

  it('should 6', async () => {
    const res = await transformWxsCode(`export function aa() {
  return new Date()
}
`)

    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })

  it('should 7', async () => {
    const res = await transformWxsCode(
      await fs.readFile(
        path.resolve(__dirname, '../fixtures/wxs/utis.wxs.ts'),
        {
          encoding: 'utf8',
        },
      ),
    )

    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })

  it('should 8', async () => {
    const res = await transformWxsCode(
      await fs.readFile(
        path.resolve(__dirname, '../fixtures/wxs/row.wxs.ts'),
        {
          encoding: 'utf8',
        },
      ),
    )

    if (res) {
      expect(res.code).toMatchSnapshot()
    }
  })
})
