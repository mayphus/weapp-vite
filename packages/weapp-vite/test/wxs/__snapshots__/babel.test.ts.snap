// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`babel > should 0 1`] = `
"module.exports.foo = module.exports.bar = void 0;
var foo = module.exports.foo = "'hello world' from comm.wxs";
var bar = module.exports.bar = function bar(d) {
  return d;
};"
`;

exports[`babel > should 1 1`] = `
"module.exports.foo = module.exports.bar = void 0;
var foo = module.exports.foo = "'hello world' from comm.wxs";
var bar = module.exports.bar = function bar(d) {
  return d;
};"
`;

exports[`babel > should 2 1`] = `
"module.exports.aa = aa;
function aa() {
  return getRegExp('{|}|"', 'g');
}"
`;

exports[`babel > should 3 1`] = `
"module.exports.bb = bb;
function bb() {
  return getRegExp("{|}|\\"", "g");
}"
`;

exports[`babel > should 4 1`] = `
"module.exports.bb = bb;
function bb() {
  return getRegExp("{|}|\\"");
}"
`;

exports[`babel > should 5 1`] = `
"module.exports.aa = aa;
function aa() {
  return getRegExp('{|}|"');
}"
`;

exports[`babel > should 6 1`] = `
"module.exports.aa = aa;
function aa() {
  return getDate();
}"
`;

exports[`babel > should 7 1`] = `
"module.exports._style = _style;
module.exports.addUnit = addUnit;
module.exports.cls = cls;
module.exports.endsWith = endsWith;
module.exports.getBadgeAriaLabel = getBadgeAriaLabel;
module.exports.includes = includes;
module.exports.isArray = isArray;
module.exports.isNoEmptyObj = isNoEmptyObj;
module.exports.isObject = isObject;
module.exports.isString = isString;
module.exports.isValidIconName = isValidIconName;
module.exports.kebabCase = kebabCase;
module.exports.keys = keys;
/* utils */

/**
 * addUnit */
// 为 css 添加单位
function addUnit(value) {
  var REGEXP = getRegExp('^-?\\d+(.\\d+)?$');
  if (value == null) {
    return undefined;
  }
  return REGEXP.test('' + value) ? value + 'px' : value;
}
function isString(string) {
  return string && string.constructor === 'String';
}
function isArray(array) {
  return array && array.constructor === 'Array';
}
function isObject(obj) {
  return obj && obj.constructor === 'Object';
}
function isNoEmptyObj(obj) {
  return isObject(obj) && JSON.stringify(obj) !== '{}';
}
;
function includes(arr, value) {
  if (!arr || !isArray(arr)) return false;
  var i = 0;
  var len = arr.length;
  for (; i < len; i++) {
    if (arr[i] === value) return true;
  }
  return false;
}
function cls(base, arr) {
  var res = [base];
  var i = 0;
  for (var size = arr.length; i < size; i++) {
    var item = arr[i];
    if (item && item.constructor === 'Array') {
      var key = arr[i][0];
      var value = arr[i][1];
      if (value) {
        res.push(base + '--' + key);
      }
    } else if (typeof item === 'string' || typeof item === 'number') {
      if (item) {
        res.push(base + '--' + item);
      }
    }
  }
  return res.join(' ');
}
function getBadgeAriaLabel(options) {
  var maxCount = options.maxCount || 99;
  if (options.dot) {
    return '有新的消息';
  }
  if (options.count === '...') {
    return '有很多消息';
  }
  if (isNaN(options.count)) {
    return options.count;
  }
  var str1 = '有' + maxCount + '+条消息';
  var str2 = '有' + options.count + '条消息';
  return Number(options.count) > maxCount ? str1 : str2;
}
function endsWith(str, endStr) {
  return str.slice(-endStr.length) === endStr ? str : str + endStr;
}
function keys(obj) {
  return JSON.stringify(obj).replace(getRegExp('{|}|"', 'g'), '').split(',').map(function (item) {
    return item.split(':')[0];
  });
}
function kebabCase(str) {
  return str.replace(getRegExp('[A-Z]', 'g'), function (ele) {
    return '-' + ele;
  }).toLowerCase();
}
function _style(styles) {
  if (isArray(styles)) {
    return styles.filter(function (item) {
      return item != null && item !== '';
    }).map(function (item) {
      return isArray(item) ? _style(item) : endsWith(item, ';');
    }).join(' ');
  }
  if (isObject(styles)) {
    return keys(styles).filter(function (key) {
      return styles[key] != null && styles[key] !== '';
    }).map(function (key) {
      return [kebabCase(key), [styles[key]]].join(':');
    }).join(';');
  }
  return styles;
}
function isValidIconName(str) {
  return getRegExp('^[A-Za-z0-9\\-]+$').test(str);
}"
`;

exports[`babel > should 8 1`] = `
"module.exports.getRowStyles = getRowStyles;
// import * as utils from './utis.wxs'
var utils = require('./utis.wxs');
function getRowStyles(gutter, style, customStyle) {
  var _style = '';
  if (gutter) {
    _style = utils._style({
      'margin-right': utils.addUnit(-gutter / 2),
      'margin-left': utils.addUnit(-gutter / 2)
    });
  }
  return utils._style([style, customStyle]) + _style;
}"
`;
